lib/lynis/lynistest.pm

```diff
@@ -0,0 +1,290 @@
+# Copyright (C) 2021 SUSE LLC
+#
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation; either version 2 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License along
+# with this program; if not, see <http://www.gnu.org/licenses/>.
+
+# Summary: Base module for Lynis test cases
+# Maintainer: llzhao <llzhao@suse.com>
+# Tags: poo#78224
+
+package lynis::lynistest;
+
+use strict;
+use warnings;
+use testapi;
+use utils;
+use autotest;
+
+use base 'consoletest';
+use bmwqemu ();
+use LTP::TestInfo 'testinfo';
+
+our @EXPORT = qw(
+  $testdir
+  $f_position_b
+  $f_position_c
+  @lynis_ok
+  @lynis_fail
+  @lynis_softfail
+  $lynis_baseline_file
+  $lynis_baseline_file_default
+  $lynis_audit_system_current_file
+  $lynis_audit_system_error_file
+  parse_lynis_section_list
+  load_lynis_section_tests
+  parse_lynis_section_content
+  compare_lynis_section_content
+);
+
+our $testdir = "/tmp/";
+
+# File pointer
+our $f_position_b = 0;
+our $f_position_c = 0;
+
+our $lynis_baseline_file_default = "baseline-lynis-audit-system-nocolors-sle15sp3-x86_64-snapshot7-textmode";
+our $lynis_baseline_file         = get_var("LYNIS_BASELINE_FILE", $lynis_baseline_file_default);
+
+our $lynis_audit_system_current_file = "lynis_audit_system_current_file";
+our $lynis_audit_system_error_file   = "lynis_audit_system_error_file";
+
+# Lynix test status mappping with openQA
+our @lynis_ok       = split(/,/, get_var("LYNIS_OK",      "OK,DONE,YES"));
+our @lynis_fail     = split(/,/, get_var("LYNIS_ERROR",   "ERROR,WEAK,UNSAFE"));
+our @lynis_softfail = split(/,/, get_var("LYNIS_WARNING", "WARNING,EXPOSED,NONE,SUGGESTION"));
+
+sub loadtest_lynis {
+    my ($test, %args) = @_;
+    autotest::loadtest("lib/lynis/$test.pm", %args);
+}
+
+# Parse lynis test report sections
+# input: $file - file name
+# output: @section_list - section list
+sub parse_lynis_section_list {
+    my ($file) = @_;
+    # Section token
+    my $s_tok        = '\[\+\] ';
+    my @section_list = ();
+
+    my $rf;
+    my $i = 0;
+
+    # Parse the "sections" of input file
+    open($rf, $file) or die "Can not open $file, $!";
```

Comment:
$! means?  If my understanding is correct, it should be $i

Reply:
The $! means $OS_ERROR, it is the return value of OS.

Reply:
Then, no question here.